### PR TITLE
eliminate possibility of double-calling

### LINKED ISSUE #3439: TestCacheExpire in ip_cache_test.go is flaky
Failed due to an apparently unrelated PR:
```
--- FAIL: TestCacheExpire (0.02s)
	ip_cache_test.go:41: Unexpected calls: [ip-address]
FAIL
coverage: 37.1% of statements
FAIL	github.com/GoogleCloudPlatform/kubernetes/pkg/master	0.138s
```
https://travis-ci.org/GoogleCloudPlatform/kubernetes/jobs/46919298

### PR DESCRIPTION
**Pull request by lavalamp**
Fixes #3439.

### CODE REVIEW

File: pkg/util/time_cache.go [hunk 1]
```diff
@@ -113,13 +119,13 @@ func (c *timeCache) fillOrWait(key string) chan T {
 
 		// Store in cache
 		c.lock.Lock()
+		defer c.lock.Unlock()
 		c.cache[key] = data
-		c.lock.Unlock()
 
 		// Remove in flight entry
 		c.inFlightLock.Lock()
```
Comment (bgrant0607) [inline pkg/util/time_cache.go]: The desired behavior is for this to block until fillOrWait returns?

Comment (lavalamp) [inline pkg/util/time_cache.go]: No, until the anonymous function called via go returns.

Comment (bgrant0607) [inline pkg/util/time_cache.go]: You deferred `c.inFlightLock.Unlock()` in fillOrWait. Therefore, the anonymous function can't continue until fillOrWait returns.

Comment (lavalamp) [inline pkg/util/time_cache.go]: Sorry, I misunderstood your question. Yes, the anonymous function can't finish until after fillOrWait returns. This is on purpose.

File: pkg/util/time_cache_test.go [hunk 2]
```diff
@@ -115,3 +117,57 @@ func TestCacheParallelOneCall(t *testing.T) {
...
+	ff := func(key string) T {
+		ct := calls[key]
+		ct.Lock()
+		ct.accessTimes = append(ct.accessTimes, time.Now())
+		ct.Unlock()
+		time.Sleep(time.Millisecond)
+		return key
+	}
+
+	clock := &FakeClock{time.Now()}
```
Comment (bgrant0607) [inline pkg/util/time_cache_test.go]: This is just preventing time from advancing in the TimeCache?

File: pkg/util/time_cache_test.go [hunk 3]
```diff
@@ -115,3 +117,57 @@ func TestCacheParallelOneCall(t *testing.T) {
...
+	c := NewTimeCache(clock, cacheDur, ff)
+
+	runtime.GOMAXPROCS(16)
+	var wg sync.WaitGroup
+	for i := 0; i < 500; i++ {
+		wg.Add(1)
+		go func(seed int64) {
+			r := rand.New(rand.NewSource(seed))
+			for i := 0; i < 50; i++ {
+				c.Get(keys[r.Intn(len(keys))])
```
Comment (bgrant0607) [inline pkg/util/time_cache_test.go]: This is to pull stuff into the cache?

A few comments would be helpful.

File: pkg/util/time_cache_test.go [hunk 4]
```diff
@@ -115,3 +117,57 @@ func TestCacheParallelOneCall(t *testing.T) {
...
+	for _, k := range keys {
+		calls[k] = &callTrack{}
+	}
+
+	ff := func(key string) T {
+		ct := calls[key]
+		ct.Lock()
+		ct.accessTimes = append(ct.accessTimes, time.Now())
+		ct.Unlock()
+		time.Sleep(time.Millisecond)
```
Comment (bgrant0607) [inline pkg/util/time_cache_test.go]: What effect is this supposed to have when using FakeClock?

File: pkg/util/time_cache_test.go (hunk 2 above)
Comment (lavalamp) [inline pkg/util/time_cache_test.go]: Oops, I meant to use RealClock for this test. Will fix.

Comment (lavalamp) [inline pkg/util/time_cache_test.go]: (btw I wrote this test before writing the fix, and it was actually catching the bug even with a non-advancing clock.)

**Comment by lavalamp** [comment]
Comments addressed, PTAL

File: pkg/util/time_cache_test.go [hunk 5]
```diff
@@ -115,3 +117,67 @@ func TestCacheParallelOneCall(t *testing.T) {
...
+	// Since the cache should hold things for 10ms, no calls for a given key
+	// should be more closely spaced than that.
+	for k, ct := range calls {
+		if len(ct.accessTimes) < 2 {
+			continue
+		}
+		cur := ct.accessTimes[0]
+		for i := 1; i < len(ct.accessTimes); i++ {
+			next := ct.accessTimes[i]
+			if next.Sub(cur) < cacheDur {
```
Comment (bgrant0607) [inline pkg/util/time_cache_test.go]: Does this really work reliably at this timescale? What about NTP setting back the clock, or TSC changes due to rescheduling to a different socket? My experience is that one can't count on clock monotonicity except at the granularity of O(seconds).

I'm willing to merge this and wait and see, but I wouldn't be surprised if it proved to be flaky.

**Comment by smarterclayton** [comment]
I want to use time cache and cache misses with a different duration than hits.  Object to me complicating time_cache, or no issue?

**Comment by bgrant0607** [comment]
@smarterclayton What specifically would you like to do?

Note that I think a cache is not really the right way to approach this. The apiserver should keep all objects in memory always (scalability will be addressed by sharding) and should serve stale data, and the data should be updated proactively periodically, rather than on demand. Some of this is discussed in #2726, IIRC.

**Comment by smarterclayton** [comment]
I ended up implementing this as a one way breaker - once read, can't be changed without restart.  No change necessary to time cache

**Pull request merged**